modules/nlu/src/backend/engine.ts

```diff
@@ -197,4 +219,48 @@ export default class Engine implements NLUEngine {
     // error handled a level highr
     return Predict(input, Engine.tools, this.predictorsByLang)
   }
+
+  private _mergeModels(previousModel: Model, trainingOuput: Model) {
+    const { artefacts: previousArtefacts } = previousModel.data
+    const { artefacts: currentArtefacts } = trainingOuput.data
+    if (!previousArtefacts || !currentArtefacts) {
+      return previousModel
+    }
+
+    const artefacts = _.merge({}, previousArtefacts, currentArtefacts)
+    const mergedModel = _.merge({}, trainingOuput, { data: { artefacts } })
+
+    // lodash merge messes up buffers objects
+    mergedModel.data.artefacts.slots_model = new Buffer(mergedModel.data.artefacts.slots_model)
+    return mergedModel
+  }
+
+  private _updateAndGetModifiedCtx = (languageCode: string, intents: NLU.IntentDefinition[], ctxs: string[]) => {
+    const ctxCache = getOrCreateCache(this.getCtxCacheId(languageCode))
```

Comment:
I'm not sure LRU cache is necessary since we already have previous model loaded in memory. Is it because you want to save compute time ? Say you want to compare ctx hash with current and previous inputs (intents)

```ts
const prevInput = this.modelsByLang[lang]?.input
const ctxToTrain = ctxs.map(ctx => {
   const prevHash = this._computeCtxHash(lang, prevInput.intents, ctx)
   const currHash = this._computeCtxHash(lang, intents, ctx)
   return prevHash !== currHash ? ctx : null
}).filter(Boolean)
```

Keeping it with your implementation would require to set your cache when you load a model which I don't think you're doing now

Reply:
You are right; I just completely forgot that we persist the inputs in the model.

Also, my implementation saves computation time only from the second time a training is called.

So yeah, I'm on this!